graphql/src/store/resolver.rs

```diff
@@ -356,144 +217,24 @@ where
-                    let entity_types = types_for_interface[&interface.name]
-                        .iter()
-                        .map(|o| o.name.clone())
-                        .collect();
-                    EntityCollection::All(entity_types)
-                }
-            };
-            let query = EntityQuery::new(subgraph_id_for_resolve_object, self.block, collection)
-                .filter(EntityFilter::Equal(String::from("id"), Value::from(id)))
-                .first(1);
-            Ok(self.store.find(query)?.into_iter().next())
-        };
-
-        let entity = if let Some(id) = id {
-            resolve_object_with_id(id)?
-        } else {
-            // Identify whether the field is derived with @derivedFrom
-            let derived_from_field = sast::get_derived_from_field(object_type, field_definition);
-            if let Some(derived_from_field) = derived_from_field {
-                // The field is derived -> build a query for the entity that might be
-                // referencing the parent object
-
-                let mut arguments = arguments.clone();
-
-                // We use first: 2 here to detect and fail if there is more than one
-                // entity that matches the `@derivedFrom`.
-                let first_arg_name = q::Name::from("first");
-                arguments.insert(&first_arg_name, q::Value::Int(q::Number::from(2)));
-
-                let skip_arg_name = q::Name::from("skip");
-                arguments.insert(&skip_arg_name, q::Value::Int(q::Number::from(0)));
-                let mut query =
-                    build_query(object_type, self.block, &arguments, types_for_interface, 2)?;
-                Self::add_filter_for_derived_field(&mut query, parent, derived_from_field);
-
-                // Find the entity or entities that reference the parent entity
-                let entities = self.store.find(query)?;
-
-                if entities.len() > 1 {
-                    return Err(QueryExecutionError::AmbiguousDerivedFromResult(
-                        field.position.clone(),
-                        field.name.to_owned(),
-                        object_type.name().to_owned(),
-                        derived_from_field.name.to_owned(),
-                    ));
-                } else {
-                    entities.into_iter().next()
-                }
-            } else {
-                match parent {
-                    Some(q::Value::Object(parent_object)) => match parent_object.get(&field.name) {
-                        Some(q::Value::String(id)) => resolve_object_with_id(id)?,
-                        _ => None,
-                    },
-                    _ => panic!("top level queries must either take an `id` or return a list"),
-                }
-            }
-        };
-
-        Ok(entity.map_or(q::Value::Null, Into::into))
+        self.resolve_object_prefetch(object_value, field, field_definition, object_type)
```

Comment:
You could inline this here -  I only made that a separate method to help readability.